### PR TITLE
Add decoding option to allow decoding CBOR byte string into Go time.Time

### PR DESCRIPTION
**Pull request by ssuriyan7**
Currently, when cbor.Unmarshal is called with a byte string data item and time.Time as destination type, it causes UnmarshalTypeError. This change adds a decoding option to specify if this operation can be allowed. 
- Default value for this option (ByteStringToTimeForbidden) blocks decoding a byte string data item into a Go time.Time value.
- ByteStringToTimeAllowed allows decoding a byte string data item into a Go time.Time value.

#### PR Was Proposed and Welcomed in Currently Open Issue

- [x] This PR was proposed and welcomed by maintainer(s) in issue #497 
- [x] Closes or Updates Issue #497 

#### Checklist (for code PR only, ignore for docs PR)

- [x] Include unit tests that cover the new code
- [x] Pass all unit tests 
- [ ] Pass all lint checks in CI (goimports, gosec, staticcheck, etc.)
- [x] Sign each commit with your real name and email.  
      Last line of each commit message should be in this format:  
      Signed-off-by: Firstname Lastname <firstname.lastname@example.com>
- [x] Certify the Developer's Certificate of Origin 1.1
      (see next section).

#### Certify the Developer's Certificate of Origin 1.1

- [x] By marking this item as completed, I certify 
      the Developer Certificate of Origin 1.1.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
660 York Street, Suite 102,
San Francisco, CA 94110 USA

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -1328,6 +1355,16 @@ func (d *decoder) parseToTime() (time.Time, bool, error) {
 	}
 
 	switch t := d.nextCBORType(); t {
+	case cborTypeByteString:
+		if d.dm.byteStringToTime == ByteStringToTimeAllowed {
+			b, _ := d.parseByteString()
+			t, err := time.Parse(time.RFC3339, string(b))
+			if err != nil {
+				return time.Time{}, false, errors.New("cbor: cannot set " + string(b) + " for time.Time: " + err.Error())
```
Comment (fxamacker) [inline decode.go:1363]: We should escape control characters and non-printable characters in the error message.

```suggestion
				return time.Time{}, false, errors.New("cbor: cannot set " + strconv.Quote(string(b)) + " for time.Time: " + err.Error())
```

File: decode_test.go [hunk 2]
```diff
@@ -8748,3 +8749,83 @@ func TestDecModeTimeTagToAny(t *testing.T) {
...
+			opts: DecOptions{ByteStringToTime: ByteStringToTimeAllowed},
+			in:   hexDecode("56323031332D30332D32315432303A30343A30302E355A"), // '2013-03-21T20:04:00.5Z'
+			want: time.Date(2013, 3, 21, 20, 4, 0, 500000000, time.UTC),
+		},
+		{
+			name:         "Unmarshal an invalid byte string to time.Time when ByteStringToTime is set to ByteStringToTimeAllowed",
+			opts:         DecOptions{ByteStringToTime: ByteStringToTimeAllowed},
+			in:           hexDecode("4B696E76616C696454657874"), // 'invalidText'
+			wantErrorMsg: "cbor: cannot set invalidText for time.Time: parsing time \"invalidText\" as \"2006-01-02T15:04:05Z07:00\": cannot parse \"invalidText\" as \"2006\"",
+		},
```
Comment (fxamacker) [inline decode_test.go:8812]: Maybe add a test with byte string containing some invalid UTF-8 characters.

**Review by fxamacker — COMMENTED** [review]
@ssuriyan7 Thanks for opening this PR!  I left a couple of suggestions.

**Comment by benluddy** [comment]
I opened https://github.com/fxamacker/cbor/pull/524 to nudge this over the finish line. This is necessary for cases that were dependent on the behavior of the bug https://github.com/fxamacker/cbor/issues/501, which has since been fixed. Thank you @ssuriyan7 for your work on this! I took care to retain you as the commit author on my updated branch.

**Comment by fxamacker** [comment]
This PR is completed by @benluddy's PR #524.  

Thanks @ssuriyan7 for opening this PR and your other contributions too!  :+1:

**Pull request closed without merging**